### PR TITLE
Custom status expiry

### PR DESCRIPTION
**Pull request by manojmalik20**
#### Summary
Added expiry support in custom statuses

#### Related Links
Specs: https://mattermost.atlassian.net/wiki/spaces/CS/pages/983302547/User+Experience+and+Functional+Notes

#### Related Pull Requests
- https://github.com/mattermost/mattermost-api-reference/pull/636
- https://github.com/mattermost/mattermost-server/pull/17570

#### Screenshots

image

image

image

image

image

image

image

image

#### Release Note
```release-note
- Added the feature to select an expiry time for custom statuses
- Made some new components like ExpiryMenu and DateTimeInput in custom status modal
- Added unit tests for new components
- Added expiry time with custom status in tooltips, status dropdown and profile popover
```

### CODE REVIEW

**Comment by esethna** [comment]
@manojmalik20 let us know when this is ready for a UX review?

**Comment by manojmalik20** [comment]
@esethna This PR is ready for a UX review.

**Comment by matthewbirtch** [comment]
@manojmalik20 I just tried testing this and it looks like setting status isn't working - at least not on the test server. I tried setting up a fresh test server to see if that would resolve it and still had the same problem. 

I open up the custom status modal, enter a status and hit the 'set status' button and it doesn't actually set the status. So I can't really test this out until that's resolved. Any ideas what the issue is?

**Comment by manojmalik20** [comment]
@matthewbirtch Yeah, I think the issue is due to the server branch. The server branch should be `custom_status_expiry`.

**Comment by Willyfrog** [comment]
Looks like the 18n tool is failing, you should run it locally (`make i18n-extract`) and fix the problem, but let us know if you have any questions regarding that

**Comment by Willyfrog** [comment]
please, remove the `package-lock.json` file from your commit as it will generate conflicts and there was no change to package.json so no need to update it.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
I still can't seem to actually get a custom status to save. However, I was able to test the interactions and there are a few things I've noticed from a UI perspective:
1. The expiry time displayed in the dropdown in the screenshot below should be `font-weight: 600` as per [the designs](https://www.figma.com/file/9OsvCq4z83xH0Vd6e68Ls9/MM-24340-Custom-Statuses?node-id=7525%3A61334). Also, the text color should have an opacity of 72% on both the 'Clear after' and the expiry time. It looks like that opacity is only applied to the 'Clear after' title.
Screen Shot 2021-05-11 at 9 51 21 AM

2. The position of the clear icon in the status input field is quite misaligned. See the screenshot above

3. I didn't I notice this before, but somewhere along the lines, the spacing/padding on the modal got broken. It's not equal spacing all around - it should be 24px all around. Can we update this in this PR? Or should we do this in a follow-up PR? It looks like it's an issue in production as well.
Screen Shot 2021-05-11 at 9 54 04 AM

4. Also, I'm seeing that the items in the suggestions/recents aren't spanning the same width as the input field itself.
Screen Shot 2021-05-11 at 10 01 36 AM

5. In the suggestions list, the expiry time should be 56% opacity as per design
6. The `Date` and `Time` labels on the custom time input fields are too big and are the wrong opacity. `font-size` on these labels should be 10px and the opacity should be 64% as per the designs.
Screen Shot 2021-05-11 at 10 06 38 AM
7. The selected date in the date picker is using an oval shape rather than a circle.
Screen Shot 2021-05-11 at 10 07 11 AM

8. The calendar and time icons aren't right in the date/time input fields. Can we update those? (these use the `compass icons` icon font.
image
image

Once I can actually get the status to save and once the above fixes are made, I can do another pass.

**Comment by manojmalik20** [comment]
> Also, I'm seeing that the items in the suggestions/recents aren't spanning the same width as the input field itself

@matthewbirtch I fixed the padding of the modal but I think the above-mentioned issue should be solved in a different PR because it can take some time as the width of the suggestion changes based on the presence/absence of the scrollbar.

Also, I noticed that in the latest [designs](https://www.figma.com/file/9OsvCq4z83xH0Vd6e68Ls9/MM-24340-Custom-Statuses?node-id=7553%3A2970), there is no design for opening modal with custom status already set with custom date/time. If the user wants to change the custom date/time and keep the emoji and text same, he cannot change it because the date and time inputs are not visible. Can you update the designs?

**Comment by matthewbirtch** [comment]
> Also, I noticed that in the latest [designs](https://www.figma.com/file/9OsvCq4z83xH0Vd6e68Ls9/MM-24340-Custom-Statuses?node-id=7553%3A2970), there is no design for opening modal with custom status already set with custom date/time. If the user wants to change the custom date/time and keep the emoji and text same, he cannot change it because the date and time inputs are not visible. Can you update the designs?

@manojmalik20 this design below should work for all set expiry times. Once a status is set, the `clear after` dropdown always shows the time when the expiry will end. If you want to change that time you open the dropdown and select 'choose data and time'

image

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: components/custom_status/date_time_input.tsx [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    if (diff === 0) {
+        return value;
+    }
+    const remainder = roundedTo - diff;
+    return start.add(remainder, 'm').seconds(0).milliseconds(0);
+}
+
+const getTimeInIntervals = (startTime: Moment): Date[] => {
+    const interval = CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES;
+    let time = moment(startTime);
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply components/custom_status/date_time_input.tsx:83]: Actually the reason for using moment here was that the native Date object doesn't store any information about the timezone. That's why when we get the current Time according to the user's timezone and convert it to native Date object, it just keeps the date and time and discards the timezone. And when we pass that time to the Timestamp component, it just renders wrong timestamp. If we cannot use moment, perhaps we can use dayjs? What do you think?

File: components/custom_status/expiry_time.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            />{' '}
+            <Timestamp
+                {...timestampProps}
+            />
+            {withinBrackets && ')'}
+        </span>
+    );
+};
+
+export default React.memo(ExpiryTime);
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply components/custom_status/expiry_time.tsx:55]: `React.memo` ensures that the component rerenders only when the props are changed.

File: selectors/views/custom_status.ts [hunk 3]
```diff
@@ -7,18 +7,24 @@ import {getConfig} from 'mattermost-redux/selectors/entities/general';
...
-        (state: GlobalState, userID?: string) => (userID ? getUser(state, userID) : getCurrentUser(state)),
-        (user) => {
-            const userProps = user?.props || {};
-            return userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
-        },
-    );
+import {getCurrentUserTimezone} from 'selectors/general';
+import {getCurrentDateTimeForTimezone} from 'utils/timezone';
+
+export function getCustomStatus(state: GlobalState, userID?: string): UserCustomStatus | undefined {
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply selectors/views/custom_status.ts:16]: So you're saying to make two separate selectors - one for custom status (memoized) and one for checking if custom status has expired (not memoized). But I don't see the benefit of making an extra selector if that's what you are saying because the amount of re-rendering will be same as the 2nd selector is not memoized and is using the same parameters. What we can do is pass customStatus in the 2nd selector like this -
```
export function isCustomStatusExpired(state: GlobalState, customStatus?: UserCustomStatus) {
    if (!customStatus || customStatus.duration === CustomStatusDuration.DONT_CLEAR) {
        return false;
    }
    
    const expiryTime = moment(customStatus.expires_at)
    const timezone = getCurrentUserTimezone(state);
    const currentTime = getCurrentMomentForTimezone(timezone)
    return currentTime.isSameOrAfter(expiryTime);
}
```

But doing this can increase the delay in expiring custom status which is already approx. 1 minute.

File: components/custom_status/date_time_input.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/custom_status/date_time_input.tsx:83]: dayjs might be nice to look into in the future, but it sounds like it's not worth it to switch away from Moment right now. We've got enough other places that we're using it for timezone-related things that it'd be a decent amount of work to remove it from everywhere, so we can just keep using it here.

File: selectors/views/custom_status.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply selectors/views/custom_status.ts:16]: I don't think that should change the delay. The only delay I see is in how long it takes `getCustomStatus` to be called after the status expires, and I'm pretty sure it would happen the same either way.

I'm not sure if I understand your suggestion, but the main thing I'm trying to solve is that `JSON.parse(userProps.customStatus)` creates a new object each time it's called, and I want to memoize that part of it. Since we need to memoize it per-caller, we'd still need a `makeGetCustomStatus`. I might have gotten ahead of myself in that suggestion though since I was also trying to ensure we didn't accidentally memoize a pre-expiry status after it had expired though.

**Comment by esethna** [comment]
@manojmalik20 can you help ping whoever is needed when this is ready for re-review?

**Comment by manojmalik20** [comment]
@hmhealey This is ready for a re-review.

**Comment by matthewbirtch** [comment]
@manojmalik20 this is breaking for me. As soon as I try to set a custom status I get a blank screen

**Comment by manojmalik20** [comment]
@matthewbirtch Can you please try again?

**Comment by manojmalik20** [comment]
@matthewbirtch Yes, actually that's because of the wrong server branch. I think the bot uses `master` branch of the server when deploying and the branch which should be used is `custom_status_expiry`. Perhaps we should wait for the [server PR](https://github.com/mattermost/mattermost-server/pull/17570) to get merged and then continue with the UX review. Till then, maybe we can continue with the dev review. @hmhealey What are your thoughts on this?

**Comment by matthewbirtch** [comment]
> @matthewbirtch Yes, actually that's because of the wrong server branch. I think the bot uses `master` branch of the server when deploying and the branch which should be used is `custom_status_expiry`. Perhaps we should wait for the [server PR](https://github.com/mattermost/mattermost-server/pull/17570) to get merged and then continue with the UX review. Till then, maybe we can continue with the dev review. @hmhealey What are your thoughts on this?

Works for me. Ping me when it's ready for another ux review and the server PR is merged

**Comment by manojmalik20** [comment]
@matthewbirtch The server PR is merged so this is ready for another UX review.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Thanks @manojmalik20. It's working well, but I have a few recommendations to make things slightly clearer.

1. For the 'Recent' statuses in the modal, it feels odd to show 'Date and Time' . Let's change the language on that to say 'Custom Date and Time'
2. Also, for the 'Recent' statuses in the modal, I don't think we need to display any subtext if the recent item was set to 'Don't Clear'. Maybe we should just hide this subtext in the recent item.
3. I noticed some strange behaviour when opening up the status modal to change a status that is already set. Steps to reproduce:
    a. When my status was set to 'In a meeting', open the status modal
    b. Focus on the status input field and hit backspace on the last character. The recents and suggestions are hidden (which is expected)
    c. Enter 'g'
    d. Notice the recents and suggestions come back for some reason.
    e. Also, the 'Set Status' button is disabled in this state.

**Comment by manojmalik20** [comment]
> d. Notice the recents and suggestions come back for some reason.
e. Also, the 'Set Status' button is disabled in this state.

@matthewbirtch The recents and suggestions come back because after entering 'g' again, the status is the same as the custom status already set. Due to the same reason, the 'Set Status' button is also disabled as it was when you opened the modal. So basically we do not want the user to be able to set the same status again.

If this is not correct, can you please explain what should be the behaviour of the recents and 'Set Status' button?
P.S. this behaviour is same as the current behaviour of recents and 'Set status' button in the `master` branch.

**Comment by manojmalik20** [comment]
@matthewbirtch This is ready for re-review. Also, @hmhealey the changes that you suggested have been made so you can also re-review again.

**Comment by matthewbirtch** [comment]
> @matthewbirtch The recents and suggestions come back because after entering 'g' again, the status is the same as the custom status already set. Due to the same reason, the 'Set Status' button is also disabled as it was when you opened the modal. So basically we do not want the user to be able to set the same status again.
> 
> If this is not correct, can you please explain what should be the behaviour of the recents and 'Set Status' button?
> P.S. this behaviour is same as the current behaviour of recents and 'Set status' button in the `master` branch.

@manojmalik20 if this is the same as the current master branch, let's leave this out of this PR. I'll make a note of it and maybe we'll look at it in another ticket

**Review by matthewbirtch — APPROVED** [review]
Looks good @manojmalik20.

FYI - I created an improvement ticket for the other issue I noted: https://mattermost.atlassian.net/browse/MM-36128

**Comment by esethna** [comment]
Ping @hmhealey @Willyfrog for review here. Thanks!

File: components/custom_status/date_time_input.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Willyfrog) [reply components/custom_status/date_time_input.tsx:83]: maybe it is a good candidate for a migration campaign?

File: components/profile_popover/profile_popover.jsx [hunk 4]
```diff
@@ -257,17 +260,18 @@ class ProfilePopover extends React.PureComponent {
...
 
-        const customStatusSet = (customStatus.text || customStatus.emoji);
+        const customStatusSet = (customStatus?.text || customStatus?.emoji) && !isCustomStatusExpired;
         const canSetCustomStatus = (user.id === currentUserId);
-        const shouldShowCustomStatus = isCustomStatusEnabled && !hideStatus && customStatus && (customStatusSet || canSetCustomStatus);
+        const shouldShowCustomStatus = isCustomStatusEnabled && !hideStatus && (customStatusSet || canSetCustomStatus);
 
         if (!shouldShowCustomStatus) {
-            return null;
+            return {};
```
Comment (Willyfrog) [inline components/profile_popover/profile_popover.jsx:270]: why did you change from null to an empty object?

File: e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_1_spec.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
```
Comment (Willyfrog) [inline e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_1_spec.js:10]: this is handled by QA
```suggestion
```

File: e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_2_spec.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
```
Comment (Willyfrog) [inline e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_2_spec.js:10]:
```suggestion
```

File: e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_3_spec.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
```
Comment (Willyfrog) [inline e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_3_spec.js:10]:
```suggestion
```

File: e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_4_spec.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
```
Comment (Willyfrog) [inline e2e/cypress/integration/custom_status/custom_status_expiry/custom_status_expiry_4_spec.js:10]:
```suggestion
```

**Review by Willyfrog — CHANGES_REQUESTED** [review]
looking good, some minor changes and a question.

File: components/profile_popover/profile_popover.jsx (hunk 4 above)
Comment (manojmalik20) [reply components/profile_popover/profile_popover.jsx:270]: So that I could use destructuring like this... 
`const {customStatusContent, expiryContent} = this.renderCustomStatus();`
without checking for null.

**Review by esethna — APPROVED** [review]

**Review by Willyfrog — APPROVED** [review]

**Comment by esethna** [comment]
@hmhealey friendly ping here for review

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: components/post_view/post_header/post_header_custom_status.tsx [hunk 9]
```diff
@@ -18,9 +18,9 @@ interface ComponentProps {
     isBot: boolean;
 }
 
+const getCustomStatus = makeGetCustomStatus();
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/post_view/post_header/post_header_custom_status.tsx:21]: This'll have to be done for each component that's using `makeGetCustomStatus`

[the remark this answers is not in the document]
Comment (manojmalik20) [reply components/post_view/post_header/post_header_custom_status.tsx:21]: @hmhealey I also don't know how to do this for functional components. I have asked a question here - https://github.com/reduxjs/reselect/discussions/498

File: components/custom_status/date_time_input.test.tsx [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        timezone: 'Australia/Sydney',
+    };
+
+    it('should match snapshot', () => {
+        const wrapper = shallow(
+            <Provider store={store}>
+                <DateTimeInput {...baseProps}/>
+            </Provider>,
+        );
+        expect(wrapper.dive()).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply components/custom_status/date_time_input.test.tsx:33]: @hmhealey Can you help me with this? I found this [issue](https://github.com/enzymejs/enzyme/issues/2202) regarding the same and I've tried all the workarounds but can't seem to resolve this.

**Review by hmhealey — APPROVED** [review]

**Comment by esethna** [comment]
cc// @ogi-m for final review and merge

**Review by ogi-m — COMMENTED** [review]
Tested various expiry times, predefined and manually selected and all looks good.

@manojmalik20 The only issue I'm seeing is related to the mobile app - once the status expires on the webapp and gets cleared, it doesn't get cleared on the mobile app where it persists indefinitely

**Comment by esethna** [comment]
@ogi-m are you okay if we open a separate ticket for that issue so we can merge this?

**Comment by ogi-m** [comment]
@esethna Sounds good to me. Opened [a ticket](https://mattermost.atlassian.net/browse/MM-36611) for it.

**Review by ogi-m — APPROVED** [review]
